compiler/rustc_infer/src/infer/canonical/query_response.rs

```diff
@@ -104,24 +106,25 @@ impl<'tcx> InferCtxt<'tcx> {
         inference_vars: CanonicalVarValues<'tcx>,
         answer: T,
         fulfill_cx: &mut dyn TraitEngine<'tcx>,
+        defining_use_anchor: DefiningAnchor,
     ) -> Result<QueryResponse<'tcx, T>, NoSolution>
     where
         T: Debug + TypeFoldable<TyCtxt<'tcx>>,
     {
         let tcx = self.tcx;
 
         // Select everything, returning errors.
-        let true_errors = fulfill_cx.select_where_possible(self);
+        let true_errors = fulfill_cx.select_where_possible(self, defining_use_anchor);
```

Comment:
provide the `defining_use_anchor` when constructing the `TraitEngine` :thinking: 

considering that different calls to `select_where_possible` try to prove the same goals it would be very weird if these different calls treat opaque types differently.

Reply:
yea 😆 I considered that, but that refactoring was more annoying than this one. I'll do a follow up PR for that

Reply:
why is that more annoying than providing the `defining_use_anchor` whenever you use the fulfillment context? '^^